Fix #7, #8, #9, #10, #11 : add classification metrics

Dhruv127 opened the pull request.

jkfitzsimons reviewed (changes requested):
Excellent work overall, I hope the comments are useful. With a few small changes you will have knocked off many of the medium PRs already! 💪  (we will need to write more or you can get points for proposing PRs yourself too :) ). Try to close out the open PRs by making the required changes before moving on though please so we can mark them as complete and give you the points!

Dhruv127 commented:
please check now , have made all the requested changes

jkfitzsimons commented on diffprivlib/Classification Metrics/metrics.accuracy_score.py:
```diff
@@ -0,0 +1,18 @@
+    y_pred = np.asarray(y_pred)
+
+    # Calculate accuracy using the regular accuracy_score function
+    accuracy = metrics.accuracy_score(y_true, y_pred)
```
This line is now redundant, we can delete.

jkfitzsimons commented on diffprivlib/Classification Metrics/metrics.auc.py:
```diff
@@ -0,0 +1,22 @@
+    y_score = np.asarray(y_score)
+
+    # Calculate AUC using the regular auc function
+    auc = metrics.auc(y_true, y_score)
```
This line is now redundant, please delete all dead code.

jkfitzsimons commented on diffprivlib/Classification Metrics/metrics.average_precision_score.py:
```diff
@@ -0,0 +1,18 @@
+    y_score = np.asarray(y_score)
+
+    # Calculate average precision using the regular function
+    avg_precision = metrics.average_precision_score(y_true, y_score)
```
This is now dead code, please delete

jkfitzsimons commented on diffprivlib/Classification Metrics/metrics.confusion_matrix.py:
```diff
@@ -0,0 +1,31 @@
+            # Perturb each element of the confusion matrix using Laplace mechanism for differential privacy
+            scale = sensitivity / epsilon
+            laplace_noise = np.random.laplace(scale=scale, size=1)
+            dp_conf_matrix[i, j] += laplace_noise + dp_mean  # Incorporate dp_mean here
```
once the dp_mean is calculated it's essentially a constant. Why are you adding it to the confusion matrix here? I understand that for each element in the confusion matrix the max contribution a y_pred value could change the results by is 1 (like counting things) so you have added the laplacian noise with appropriate scale (good job), but then you add the dp_mean... what is the logic behind this? I think it's incorrect

jkfitzsimons commented on diffprivlib/Classification Metrics/metrics.f1_score.py:
```diff
@@ -0,0 +1,25 @@
+    laplace_noise = np.random.laplace(scale=scale, size=1)
+
+    # Adjust the F1 score using dp_mean and Laplace noise
+    dp_f1 = f1 + dp_mean + laplace_noise  # You can adjust the combination based on your requirements
```
This equation once again is incorrect. You are adding laplace noise directly but the sensitivity is wrong and then you are adding the dp_mean? A hint would be to get the dp confusion matrix working. Once you have that the f1 function can simply call it, get the result and use the TP, FP, FN from the result to calculate f1-score directly. Check the wikipedia for their relationship: https://en.wikipedia.org/wiki/F-score

jkfitzsimons reviewed (changes requested):
There are some fundamental errors in the code which I've marked up. Not sure if you are using GPT or something on the confusion matrix and f1 scores but I'm certain it's incorrect. Hopefully the comments guide you to the correct solution :) Happy to answer questions if you get stuck, you are almost there!

Dhruv127 commented:
bhaiya , please check now , and don't i know if its allowed but i use the ai for some reference and to add comments , sorry for taking your so much time.
Please review it once more.

jkfitzsimons commented on diffprivlib/Classification_Metrics/average_precision_score.py:
```diff
@@ -0,0 +1,15 @@
+    accountant = BudgetAccountant()
+    dp_mean = mean(y_score, bounds=bounds, epsilon=epsilon, random_state=random_state, accountant=accountant)
+
+    return avg_precision + dp_mean
```
What is going on here - you are getting the regular precision and then adding the dp_mean of the y_score..? How would this help? Seen as you have the confusion matrix already built, why not use it directly, see: https://towardsdatascience.com/accuracy-precision-recall-or-f1-331fb37c5cb9

jkfitzsimons reviewed (changes requested):
I think there is one remaining bug. I hope I've been very clear how you can solve it :)

Dhruv127 commented:
@jkfitzsimons , is it done ?

Dhruv127 commented:
@jkfitzsimons  please check now , i have updated the confusion matrix . And for average_precision  , i can't find any direct method to calculate it from recall and precision , so i added Laplace noise , to the one obtained from sklearn.average_precison method . Please check it once .

jkfitzsimons commented on diffprivlib/Classification_Metrics/average_precision_score.py:
```diff
@@ -0,0 +1,29 @@
+
+    output = base_lap(output)
+
+    return output
```
What you are doing here is adding noise to precision as if it was count data (1,2,3,4,5) but the precision is in (0-1). You might be better off calculating the conf_matrix = metrics.confusion_matrix(y_true, y_pred), and then adding noise to TP and FP as you have (clip to 0, ie no value should be negative), and the precision = TP/(TP+FP) as in https://machinelearningmastery.com/precision-recall-and-f-measure-for-imbalanced-classification/#:~:text=Consider%20a%20model%20that%20predicts,95%20%2F%20(95%20%2B%2055)

jkfitzsimons reviewed (changes requested):
Last change request hopefully :)
